### PR TITLE
#43 - Associate users with roles

### LINKED ISSUE #43: Associate users with roles
It will allow us to easily authorize actions and block routes.

### PR DESCRIPTION
**Pull request by PiotrFedak**
This should close #43

### CODE REVIEW

**Comment by PiotrFedak** [comment]
In my console all test are passed 
Screenshot_20230627_111907

I don't get it

**Comment by EwelinaLasowy** [comment]
Failed tests.

File: resources/js/Pages/Admin.vue [hunk 1]
```diff
@@ -0,0 +1,11 @@
+<template>
+  <div class="flex items-center bg-blue-100 p-2">
+    <div>
+      <h1>Page for ADMIN ONLY</h1>
+      <a href="/" class="absolute right-4 top-4 text-blue-600 underline">Go Back</a>
+    </div>
+    <span class="px-4" />
+  </div>
+</template>
+
```
Comment (EwelinaLasowy) [inline resources/js/Pages/Admin.vue:10]: We need only one empty line at the end of the file.

File: routes/web.php [hunk 2]
```diff
@@ -23,10 +23,14 @@
 Route::middleware("auth")->group(function (): void {
     Route::post("/logout", [LogoutController::class, "logout"])->name("logout");
     Route::get("/dashboard", fn(): Response => inertia("Dashboard"))->name("dashboard");
+
+    Route::middleware(["role:admin"])->group(function (): void {
```
Comment (EwelinaLasowy) [inline routes/web.php:27]: I'm not sure if this middleware "role:admin" should be like this?

Comment (vojcc) [reply routes/web.php:27]: Why not?

Comment (EwelinaLasowy) [reply routes/web.php:27]: I mean that this middleware for now it's okay. But when you will have permissions for each action, it should change.

Comment (EwelinaLasowy) [reply routes/web.php:27]: We generally use gates/policy for stuff like this and not on middleware level. Also you can use permissions.

File: app/Http/Kernel.php [hunk 3]
```diff
@@ -87,5 +87,6 @@ class Kernel extends HttpKernel
         "signed" => ValidateSignature::class,
         "throttle" => ThrottleRequests::class,
         "verified" => EnsureEmailIsVerified::class,
+        "role" => \Spatie\Permission\Middlewares\RoleMiddleware::class,
```
Comment (kamilpiech97) [inline app/Http/Kernel.php:90]: Please import that class

File: app/Models/User.php [hunk 4]
```diff
@@ -45,4 +48,9 @@ class User extends Authenticatable
         "email_verified_at" => "datetime",
         "password" => "hashed",
     ];
+
+    public function Admin()
```
Comment (kamilpiech97) [inline app/Models/User.php:52]:
```suggestion
    public function isAdmin()
```

File: database/seeders/RoleSeeder.php [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+namespace Database\Seeders;
+
+use Illuminate\Database\Seeder;
+use Spatie\Permission\Models\Role;
+
+class RoleSeeder extends Seeder
+{
+    /**
+     * Run the database seeds.
+     */
```
Comment (kamilpiech97) [inline database/seeders/RoleSeeder.php:14]:
```suggestion
```

File: database/seeders/AdminSeeder.php [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    {
+        $password = Str::random(10);
+
+        User::create([
+            "name" => "admin",
+            "email" => "admin@example.com",
+            "email_verified_at" => now(),
+            "password" => Hash::make($password),
+            "remember_token" => Str::random(10),
+        ])->assignRole("admin");
```
Comment (kamilpiech97) [inline database/seeders/AdminSeeder.php:27]:
```suggestion
        User::create([
            "name" => "admin",
            "email" => "admin@example.com",
            "email_verified_at" => now(),
            "password" => Hash::make("password"),
            "remember_token" => Str::random(10),
        ])->assignRole("admin");
```

Comment (kamilpiech97) [reply database/seeders/AdminSeeder.php:27]: I think it will be good to know what password the user has.

File: app/Models/User.php [hunk 7]
```diff
@@ -45,4 +48,9 @@ class User extends Authenticatable
         "email_verified_at" => "datetime",
         "password" => "hashed",
     ];
+
+    public function isAdmin()
+    {
+        return $this->role === "admin";
+    }
```
Comment (kamilpiech97) [inline app/Models/User.php:55]:
```suggestion
    public function isAdmin(): bool
    {
        return $this->hasRole("admin");
    }
```

File: database/migrations/2023_06_27_000000_add_role_to_users_table.php [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+        });
+    }
+
+    public function down(): void
+    {
+        Schema::table("users", function (Blueprint $table): void {
+            $table->dropColumn("role");
+        });
+    }
+}
```
Comment (kamilpiech97) [inline database/migrations/2023_06_27_000000_add_role_to_users_table.php:24]: Do we need that field in `User` model?

Comment (kamilpiech97) [reply database/migrations/2023_06_27_000000_add_role_to_users_table.php:24]: In factory we are not using that field.

**Review by kamilpiech97 — CHANGES_REQUESTED** [review]

**Review by kamilpiech97 — APPROVED** [review]

**Review by EwelinaLasowy — APPROVED** [review]

**Review by vojcc — APPROVED** [review]

**Pull request merged**